source/common/upstream/eds.cc

```diff
@@ -144,7 +186,62 @@ void EdsClusterImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef
     assignment_timeout_->enableTimer(std::chrono::milliseconds(stale_after_ms));
   }
 
-  BatchUpdateHelper helper(*this, cluster_load_assignment);
+  // Pause LEDS messages until the EDS config is finished processing.
+  Config::ScopedResume maybe_resume_leds;
+  if (factory_context_.clusterManager().adsMux()) {
+    const auto type_url = Config::getTypeUrl<envoy::config::endpoint::v3::LbEndpoint>();
+    maybe_resume_leds = factory_context_.clusterManager().adsMux()->pause(type_url);
+  }
+
+  // Compare the current set of LEDS localities (localities using LEDS) to the one received in the
+  // update. A LEDS locality can either be added, removed, or kept. If it is added we add a
+  // subscription to it, and if it is removed we delete the subscription.
+  LedsConfigSet cla_leds_configs;
+
+  for (const auto& locality : cluster_load_assignment.endpoints()) {
+    if (locality.has_leds_cluster_locality_config()) {
+      cla_leds_configs.emplace(locality.leds_cluster_locality_config());
+    }
+  }
+
+  // Remove the LEDS localities that are not needed anymore.
+  absl::erase_if(leds_localities_, [&cla_leds_configs](const auto& item) {
+    auto const& [leds_config, _] = item;
+    // Returns true if the leds_config isn't in the cla_leds_configs
+    return cla_leds_configs.find(leds_config) == cla_leds_configs.end();
+  });
+
+  // Add all the LEDS localities that are new.
+  for (const auto& leds_config : cla_leds_configs) {
+    if (leds_localities_.find(leds_config) == leds_localities_.end()) {
+      ENVOY_LOG(debug, "Found new LEDS config in EDS onConfigUpdate() for cluster {}: {}",
+                cluster_name_, leds_config.DebugString());
+
+      // Create a new LEDS subscription and add it to the subscriptions map.
+      LedsSubscriptionPtr leds_locality_subscription = std::make_unique<LedsSubscription>(
+          leds_config, cluster_name_, factory_context_, info_->statsScope(), [&]() {
+            // Called upon an update to the locality.
+            if (validateAllLedsUpdated()) {
+              BatchUpdateHelper helper(*this);
+              priority_set_.batchHostUpdate(helper);
```

Comment:
It is called once per LEDS update, and should cover the following cases:
1. Before all the locality endpoints of a specific EDS cluster are ready (this will fail the `validateAllLedsUpdated()` call, and no actual update will occur.
2. Whenever a new update for some locality arrives.